src/librustc_typeck/astconv.rs

```diff
@@ -919,6 +903,22 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
         })
         .expect("missing associated type");
 
+        // Make sure the supplied parameters match the declared parameter of
+        // the associated type.
+        let (ty_provided, lt_provided) = match &item_segment.parameters {
+            Some(params) => (params.types.len(), params.lifetimes.len()),
+            None => (0, 0),
+        };
+
+        let (ty_params, lt_accepted)
+            = accepted_argument_count_of(tcx.generics_of(item.def_id));
+
+        if lt_accepted != lt_provided {
+            report_lifetime_number_error(tcx, span, lt_provided, lt_accepted);
+        }
+        check_type_argument_count(tcx, span, ty_provided, ty_params);
+
+
         let ty = self.projected_ty_from_poly_trait_ref(span, item.def_id, bound);
```

Comment:
I think I see what's missing here - `projected_ty_from_poly_trait_ref` should not exist. `create_substs_for_ast_path` should be used, and it now needs to handle multiple segments containing parameters, which is unprecedenced.